# Advent of Code 2020 Day 1: Report Repair

## Import libraries

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None)

## Part 1

### Description Part 1

After saving Christmas five years in a row, you've decided to take a vacation at a nice resort on a tropical island. Surely, Christmas will go on without you.

The tropical island has its own currency and is entirely cash-only. The gold coins used there have a little picture of a starfish; the locals just call them stars. None of the currency exchanges seem to have heard of them, but somehow, you'll need to find fifty of these coins by the time you arrive so you can pay the deposit on your room.

To save your vacation, you need to get all fifty stars by December 25th.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

Before you leave, the Elves in accounting just need you to fix your expense report (your puzzle input); apparently, something isn't quite adding up.

Specifically, they need you to find the two entries that sum to 2020 and then multiply those two numbers together.

For example, suppose your expense report contained the following:

1721
979
366
299
675
1456

In this list, the two entries that sum to 2020 are 1721 and 299. Multiplying them together produces 1721 * 299 = 514579, so the correct answer is 514579.

Of course, your expense report is much larger. Find the two entries that sum to 2020; what do you get if you multiply them together?

### Import data

In [2]:
dfa1 = pd.read_csv('day_1_input.csv')

In [3]:
dfa1.head()

number
0    1844
1    1123
2    1490
3    1478
4    1108

### Create column with 2020 - number

In [4]:
dfa1['difference'] = 2020 - dfa1['number']

In [5]:
dfa1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   number      200 non-null    int64
 1   difference  200 non-null    int64
dtypes: int64(2)
memory usage: 3.3 KB


In [6]:
dfa1.head()

number  difference
0    1844         176
1    1123         897
2    1490         530
3    1478         542
4    1108         912

### Check if entries in difference match any entries in number

In [7]:
dfa2 = dfa1[dfa1['difference'].isin(dfa1['number'])]

In [8]:
dfa2

number  difference
34      976        1044
159    1044         976

### Multiply together for output

In [9]:
dfa2['output'] = dfa2['number'] * dfa2['difference']

C:\Users\11dmc\AppData\Local\Temp\ipykernel_52936\4122726338.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfa2['output'] = dfa2['number'] * dfa2['difference']


In [10]:
dfa2

number  difference   output
34      976        1044  1018944
159    1044         976  1018944

## Solution: The product of the two entries that sum to 2020 is 1018944

## Part 2

### Description Part 2

The Elves in accounting are thankful for your help; one of them even offers you a starfish coin they had left over from a past vacation. They offer you a second one if you can find three numbers in your expense report that meet the same criteria.

Using the above example again, the three entries that sum to 2020 are 979, 366, and 675. Multiplying them together produces the answer, 241861950.

In your expense report, what is the product of the three entries that sum to 2020?

In [11]:
# Sort values by number
dfa3 = dfa1.sort_values('number').reset_index(drop = True)

In [12]:
# Convert columns to series
number_array = np.array(dfa3['number'])

In [13]:
number_array

array([  16,  312,  976, 1005, 1014, 1016, 1024, 1033, 1035, 1040, 1044,
       1049, 1058, 1069, 1084, 1088, 1093, 1094, 1098, 1100, 1101, 1102,
       1103, 1105, 1107, 1108, 1120, 1123, 1129, 1131, 1133, 1142, 1146,
       1149, 1159, 1162, 1163, 1164, 1171, 1175, 1179, 1180, 1181, 1188,
       1196, 1201, 1215, 1219, 1220, 1224, 1225, 1250, 1253, 1260, 1267,
       1269, 1273, 1276, 1283, 1290, 1291, 1306, 1311, 1319, 1327, 1330,
       1338, 1351, 1352, 1362, 1364, 1366, 1372, 1388, 1400, 1406, 1409,
       1413, 1414, 1421, 1431, 1432, 1435, 1439, 1443, 1457, 1462, 1463,
       1470, 1472, 1474, 1477, 1478, 1482, 1488, 1490, 1493, 1498, 1499,
       1505, 1508, 1510, 1514, 1517, 1523, 1527, 1531, 1532, 1535, 1537,
       1539, 1544, 1550, 1551, 1558, 1576, 1581, 1588, 1594, 1595, 1596,
       1597, 1600, 1601, 1604, 1607, 1613, 1614, 1619, 1642, 1646, 1647,
       1653, 1658, 1659, 1668, 1678, 1683, 1685, 1689, 1692, 1697, 1698,
       1702, 1703, 1704, 1713, 1717, 1719, 1722, 17

In [14]:
# Define function to get subtract numbers from 2020, then iterate through list subtracting each entry, then check number list for matches

def triple_sum(data_array, target_sum):
 
    for i in range(len(data_array)):
        for j in range(i + 1, len(data_array)):
            for k in range(j + 1, len(data_array)):
                if data_array[i] + data_array[j] + data_array[k] == target_sum:
                    
                    return [data_array[i], data_array[j], data_array[k]]

In [15]:
triple_sum(number_array, 2020)

[16, 312, 1692]

In [16]:
16 + 312 + 1692

2020

In [17]:
16 * 312 * 1692

8446464

## Solution: The product of the 3 entries that sum to 2020 is 8446464